### ISSUE TITLE
SACHA - compilation issue AimsConnectedComponent

### ISSUE THREAD

**Issue by manuegrx** [brainvisa/aims-free#56]
Hi @denisri , 

I have compilation issues with the project "sacha-private" and they seem related to Aims: 

``` 
[ 93%] Building CXX object build_files/sacha-private/CMakeFiles/SachaSegS2.dir/hippoaims/precalc.cc.o
/casa/host/src/sacha-private/hippoaims/precalc.cc: In member function ‘void PreCalculs::doCompoHcAm()’:
/casa/host/src/sacha-private/hippoaims/precalc.cc:749:66: error: no matching function for call to ‘AimsConnectedComponent(AimsData<short int>&, aims::Connectivity::Type&, const int16_t&, bool, int, int, bool)’
  AimsConnectedComponent(*maskDataH,connexite26,BCG,true,0,0,false);
                                                                  ^
In file included from /casa/host/build/include/aims/connectivity/connectivity_g.h:50:0,
                 from /casa/host/build/include/aims/morphology/operatormorpho.h:39,
                 from /casa/host/src/sacha-private/hippoaims/precalc.cc:28:
/casa/host/build/include/aims/connectivity/component.h:125:8: note: candidate: template<class T> void aims::AimsConnectedComponent(carto::VolumeRef<T>, aims::Connectivity::Type, std::map<T, long unsigned int>&, const T&, bool, size_t, size_t, size_t, bool)
   void AimsConnectedComponent( carto::VolumeRef<T> data,
        ^~~~~~~~~~~~~~~~~~~~~~
/casa/host/build/include/aims/connectivity/component.h:125:8: note:   template argument deduction/substitution failed:
/casa/host/src/sacha-private/hippoaims/precalc.cc:749:66: note:   ‘AimsData<short int>’ is not derived from ‘carto::VolumeRef<T>’
  AimsConnectedComponent(*maskDataH,connexite26,BCG,true,0,0,false);
````

There are like 15 mistakes and they seem all related to `AimsConnectedComponent`. 
Is it because `AimsConnectedComponent` is depreciated ?

**Comment by denisri** [comment]
Yes that's "normal". I changed recently the API and implementation of connected components to use the `Volume` (`VolumeRef` actually, which is a reference-counting pointer to a `Volume`) structure instead of the antique `AimsData` classes. Apart from that the API is the same, and there are automatic casts in C++ between `AimsData` and `VolumeRef`, so things should be transparent. However... `AimsData` and `VolumeRef` are template classes (`VolumeRef<short>` for instance) and here they are used in a template function (`AimsConnectedComponent<T>`), and the C++ language has a limitation in this precise specific situation, where it cannot find by itself the `AimsData<T>` -> `VolumeRef<T>` correspondance when `T` can vary according to a template function which we do not specify which `T` it is. I spent some time searching if it was possible to have this automatic conversion, and finally found out that no, it is not possible. So the sources of Sacha have to be slightly modified. The problem can be avoided by specifying the template type of the function call, by replacing:
```
  AimsConnectedComponent(*maskDataH,connexite26,BCG,true,0,0,false);
```
with:
```
  AimsConnectedComponent<int16_t>(*maskDataH,connexite26,BCG,true,0,0,false);
```
(here I assume the voxel type of the mask is `int16_t` but use the right one as in the declaration of `maskDataH`)

I have changed this in all codes I have under my hand, but Sacha is one of the few projects I don't have.

A better solution would be to use `VolumeRef` structures in Sacha codes instead of `AimsData`.
